runtime/topk.cu

```diff
@@ -0,0 +1,166 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <cub/block/block_radix_sort.cuh>
+
+namespace nvf {
+namespace topk {
+
+// Block state constants following nvFuser conventions from fused_reduction.cu
+// Sort Domain - TEMPLATE STATE 0
+//   - Participating in the topk selection, has values coming in, selected
+//   values coming out
+// Iteration Domain - TEMPLATE STATE 1
+//   - Not participating in the topk selection, has values across the dimension
+//   after selection
+constexpr __device__ bool isSort(int STATE) {
+  return STATE == 0;
+}
+
+constexpr __device__ bool isIter(int STATE) {
+  return STATE == 1;
+}
+
+// TODO: We have exactly the same code in argsort.cu, we should refactor it.
+// Type utils for interoperability between our own half types and the
+// CUDA standard types (identical to argsort implementation)
+template <typename T>
+struct CudaType {
+  using type = T;
+
+  __device__ inline static T get(const T& t) {
+    return t;
+  }
+};
+
+#ifdef __NVFUSER_HAS_HALF__
+template <>
+struct CudaType<__half> {
+  using type = __nv_half;
+
+  __device__ inline static typename CudaType<__half>::type get(
+      const __half& t) {
+    return __ushort_as_half(__NVFUSER_HALF_TO_CUS(t));
+  }
+};
+#endif // __NVFUSER_HAS_HALF__
+
+#ifdef __NVFUSER_HAS_BFLOAT__
+template <>
+struct CudaType<__bfloat> {
+  using type = __nv_bfloat16;
+
+  __device__ inline static typename CudaType<__bfloat>::type get(
+      const __bfloat& t) {
+    return __ushort_as_bfloat16(__NVFUSER_BFLOAT_TO_CUS(t));
+  }
+};
+#endif // __NVFUSER_HAS_BFLOAT__
+
+// Block-parallel topk using CUB BlockRadixSort
+// Following nvFuser dimensional template parameter pattern like
+// fused_reduction.cu and argsort.cu
+//
+// For simplicity, we assume that:
+// - top_values: Output array containing ITEMS_PER_THREAD top-K values per
+// thread
+// - top_indices: Output array containing original indices of top-K values
+// - Each thread holds exactly ITEMS_PER_THREAD elements in both output arrays
+// - In the actual nvFuser-generated kernel, only the first k elements of each
+//   thread's output arrays should be used (via predication)
```

Comment:
> // only the first k elements of each thread's output arrays should be used (via predication)

When we do topk on blocks. I think the result here should be that "only the first k elements of all thread's combined output array" instead of first k elements of **each** thread

Reply:
You're right. Updated.